### PR TITLE
Add checkpointing and reloading models

### PR DESCRIPTION
**Pull request by edwardjkim**
*Issue #, if available:*

*Description of changes:*

Support checkpointing.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 1]
```diff
@@ -156,14 +163,28 @@ def train_job(train_cfg, train_dmatrix, val_dmatrix, model_dir, is_master):
     if val_dmatrix is not None:
         watchlist.append((val_dmatrix, 'validation'))
 
+    xgb_model, iteration = train_utils.load_checkpoint(checkpoint_dir)
+    if not is_master:
```
Comment (iyerr3) [inline src/sagemaker_xgboost_container/algorithm_mode/train.py]: Why is this necessary? Alternatively, when does the model get synced between hosts for the resumed job?

**Comment by edwardjkim** [comment]
Since the repo became public and my fork is still private, I lost cross-repo sync of PR. I made some changes in my fork, but it's not showing up here. Right now I can't make my fork public because of a branch that needs to stay private, but I think I might just need to delete that branch because making separate PRs for every commit is going to be painful for everyone involved.

File: src/sagemaker_xgboost_container/algorithm_mode/callback.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    We print the evaluation results every **period** iterations
+    and on the first and the last iterations.
+
+    Parameters
+    ----------
+    period : int
+        The period to log the evaluation results
+
+    show_stdv : bool, optional
+         Whether show stdv if provided
```
[the remark this answers is not in the document]
Comment (edwardjkim) [reply src/sagemaker_xgboost_container/algorithm_mode/callback.py]: `print_evaluation()` is a straight copy of the same function from upstream, and I added just the minimum (`start_iteration`) so that the CW logs show the right iteration number when resuming from checkpoint. I could add my understanding of what that the argument does, but I don't want to change much from upstream. It's just a comment so I guess I can add something so just let me know and I'll add something.

File: src/sagemaker_xgboost_container/algorithm_mode/callback.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+    Returns
+    -------
+    callback : function
+        A callback that print evaluation every period iterations.
+    """
+    def callback(env):
+        """internal function"""
```
[the remark this answers is not in the document]
Comment (edwardjkim) [reply src/sagemaker_xgboost_container/algorithm_mode/callback.py]: Will add in the next PR, which will be a separate PR because I don't have cross-repository auto-update of this PR as I commented in the general comment. Adding last-minute comments before closing this PR.

File: src/sagemaker_xgboost_container/algorithm_mode/callback.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        if i % period == 0 or i + 1 == env.begin_iteration or i + 1 == env.end_iteration:
+            msg = '\t'.join([fmt_metric(x, show_stdv) for x in env.evaluation_result_list])
+            rabit.tracker_print('[%d]\t%s\n' % (i + start_iteration, msg))
+    return callback
+
+
+class SaveCheckpoint:
+    """Create a callback that saves checkpoints to disk."""
+
+    SENTINEL = None
```
[the remark this answers is not in the document]
Comment (edwardjkim) [reply src/sagemaker_xgboost_container/algorithm_mode/callback.py]: Will be added in next PR.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py (hunk 1 above)
Comment (edwardjkim) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py]: When Spot resumes a host, it downloads the checkpoints to `/opt/ml/checkpoints` **before** it starts a fresh training job. Line 166 checks the checkpoint directory to see if there are any checkpoint files. If there are, it means the training was interrupted previously so we load from the latest checkpoint. If not, we start from iteration 0.

**Comment by edwardjkim** [comment]
As commented above, I need to continue this PR in a separate PR due to inability to sync between a public repo and a private fork. Closing.
